src/ir/match.h

```diff
@@ -0,0 +1,782 @@
+  return Internal::ConstMatcher(
+    nullptr, Internal::IntLit(nullptr, Internal::Any<int64_t>(nullptr)));
+}
+decltype(auto) ival(int64_t x) {
+  return Internal::ConstMatcher(
+    nullptr, Internal::IntLit(nullptr, Internal::Exact<int64_t>(nullptr, x)));
+}
+decltype(auto) ival(int x) { return ival(int64_t(x)); } // disambiguate 0
+decltype(auto) ival(int64_t* binder) {
+  return Internal::ConstMatcher(
+    nullptr, Internal::IntLit(nullptr, Internal::Any(binder)));
+}
+decltype(auto) ival(Literal* binder) {
+  return Internal::ConstMatcher(
+    nullptr, Internal::IntLit(binder, Internal::Any<int64_t>(nullptr)));
+}
+decltype(auto) ival(Const** binder) {
+  return Internal::ConstMatcher(
+    binder, Internal::IntLit(nullptr, Internal::Any<int64_t>(nullptr)));
+}
+decltype(auto) ival(Literal* binder, int64_t x) {
+  return Internal::ConstMatcher(
+    nullptr, Internal::IntLit(binder, Internal::Exact<int64_t>(nullptr, x)));
+}
+decltype(auto) ival(Const** binder, int64_t x) {
+  return Internal::ConstMatcher(
+    binder, Internal::IntLit(nullptr, Internal::Exact<int64_t>(nullptr, x)));
+}
+
+decltype(auto) fval() {
+  return Internal::ConstMatcher(
+    nullptr, Internal::FloatLit(nullptr, Internal::Any<double>(nullptr)));
+}
+decltype(auto) fval(double x) {
+  return Internal::ConstMatcher(
+    nullptr, Internal::FloatLit(nullptr, Internal::Exact<double>(nullptr, x)));
+}
+decltype(auto) fval(int x) { return fval(double(x)); } // disambiguate literal 0
+decltype(auto) fval(double* binder) {
+  return Internal::ConstMatcher(
+    nullptr, Internal::FloatLit(nullptr, Internal::Any(binder)));
+}
+decltype(auto) fval(Literal* binder) {
+  return Internal::ConstMatcher(
+    nullptr, Internal::FloatLit(binder, Internal::Any<double>(nullptr)));
+}
+decltype(auto) fval(Const** binder) {
+  return Internal::ConstMatcher(
+    binder, Internal::FloatLit(nullptr, Internal::Any<double>(nullptr)));
+}
+decltype(auto) fval(Literal* binder, double x) {
+  return Internal::ConstMatcher(
+    nullptr, Internal::FloatLit(binder, Internal::Exact<double>(nullptr, x)));
+}
+decltype(auto) fval(Const** binder, double x) {
+  return Internal::ConstMatcher(
+    binder, Internal::FloatLit(nullptr, Internal::Exact<double>(nullptr, x)));
+}
+
+decltype(auto) constant() {
+  return Internal::ConstMatcher(nullptr, Internal::Any<Literal>(nullptr));
+}
+decltype(auto) constant(int x) {
+  return Internal::ConstMatcher(nullptr, Internal::NumberLit(nullptr, x));
+}
+decltype(auto) constant(Literal* binder) {
+  return Internal::ConstMatcher(nullptr, Internal::Any(binder));
+}
+decltype(auto) constant(Const** binder) {
+  return Internal::ConstMatcher(binder, Internal::Any<Literal>(nullptr));
+}
+decltype(auto) constant(Literal* binder, int32_t x) {
+  return Internal::ConstMatcher(nullptr, Internal::NumberLit(binder, x));
+}
+decltype(auto) constant(Const** binder, int32_t x) {
+  return Internal::ConstMatcher(binder, Internal::NumberLit(nullptr, x));
+}
```

Comment:
`constant` supports all of `i32`, `i64`, `f32`, and `f64`, but that means it only matches integer float values like `1.0` or `2.0`. Since it matches only integer floats, we could in theory allow all matched values to be bound to an int32_t, but that might give readers the mistaken impression that only `i32` and `i64` constants can be matched. Users can extract the `Literal`, though, which gives them access to both the value (if they didn't already specify it) and its type.